[ResizeObserver] The initial last reported size of ResizeObservation should be -1x-1 which always notify observed elements

cathiechen opened the pull request:
#### 2861d8c4e29ca8c3384e47f260a0d0e1823c29ae
<pre>
[ResizeObserver] The initial last reported size of ResizeObservation should be -1x-1 which always notify observed elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=251015">https://bugs.webkit.org/show_bug.cgi?id=251015</a>

Reviewed by Oriol Brufau and Ryosuke Niwa.

This patch initializes the lastReportedSize of ResizeObservation to -1 x -1, which is the conclusion of [1],
it indicates that -1 x -1 is the initial size when first observing an element with ResizeObserver.

The previous change makes every ResizeObservation be delivered at least once, even if the target is a disconnected element.
When GC happens between observe and gather ResizeObservations, JSC::Weak&lt;JSC::JSObject&gt; m_callback inside JSCallbackDataWeak
could be released, and the element and JSResizeObserver might not be released, for they are not strong connected, this scenario
would trigger the ASSERT in ResizeObserver::deliverObservations. To fix it, this patch adds m_targetsWaitingForFirstObservation
to extend the life cycle of ResizeObserver to the first time of deliverObservations. The fix is similar to
what IntersectionObserver does in bug 231235.

The test6 of resize-observer/notify.html still fails in mac-wk1, because updaterendering is not triggered properly.
Filed [2] to track it.

[1] <a href="https://github.com/w3c/csswg-drafts/issues/3664">https://github.com/w3c/csswg-drafts/issues/3664</a>
[2] <a href="https://bugs.webkit.org/show_bug.cgi?id=250900">https://bugs.webkit.org/show_bug.cgi?id=250900</a>

* LayoutTests/imported/w3c/web-platform-tests/resize-observer/notify-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/resize-observer/observe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/resize-observer/svg-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/resize-observer/notify-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/resize-observer/notify-expected.txt.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/resize-observer/svg-expected.txt:
* Source/WebCore/page/ResizeObservation.cpp:
* Source/WebCore/page/ResizeObserver.cpp:
(WebCore::ResizeObserver::observe):
(WebCore::ResizeObserver::deliverObservations):
(WebCore::ResizeObserver::isReachableFromOpaqueRoots const):
(WebCore::ResizeObserver::removeAllTargets):
(WebCore::ResizeObserver::removeObservation):
* Source/WebCore/page/ResizeObserver.h:

Canonical link: <a href="https://commits.webkit.org/259673@main">https://commits.webkit.org/259673@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1408ba697584028bc6a0c5f438e923f2ecd2fe5f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105519 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14617 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38424 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114776 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174924 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109419 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16033 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5843 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97821 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114622 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111276 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12189 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95181 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39684 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94053 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26817 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81379 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7898 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28173 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8248 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4761 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14025 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47716 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6683 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9923 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

Loirooriol commented on Source/WebCore/page/ResizeObserver.cpp:
```diff
@@ -82,6 +82,8 @@ void ResizeObserver::observe(Element& target, const ResizeObserverOptions& optio
 
     m_observations.append(ResizeObservation::create(target, options.box));
 
+    m_targetsWaitingForFirstObservation.append(target);
```
Maybe add the same comment as for IntersectionObserver

```
    // Per the specification, we should dispatch at least one observation for the target. For this reason, we make sure to keep the
    // target alive until this first observation. This, in turn, will keep the ResizeObserver's JS wrapper alive via
    // isReachableFromOpaqueRoots(), so the callback stays alive.
```

cathiechen replied:
Done, thanks!

Loirooriol commented on Source/WebCore/page/ResizeObserver.cpp:
```diff
@@ -141,6 +143,8 @@ void ResizeObserver::deliverObservations()
     m_activeObservations.clear();
     auto activeObservationTargets = std::exchange(m_activeObservationTargets, { });
 
+    std::exchange(m_targetsWaitingForFirstObservation, { });
```
I'm not an expert in these things, but my worry is that, unlike `IntersectionObserver::notify`, you might not be keeping the old value alive by storing it into a variable.

Could it then be possible for the callbacks to be garbage collected before `m_callback->handleEvent(*this, entries, *this);`? I will defer the review to someone who knows better about these things.

rniwa replied:
Yeah, we should clear this after calling m_callback->handleEvent. Maybe use ScopeExit?

cathiechen replied:
Done, thanks!

Loirooriol commented on LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/resize-observer/notify-expected.txt:
```diff
@@ -0,0 +1,23 @@
+PASS test3: dimensions match
+PASS test4: transform do not cause notifications
+PASS test5: moving an element does not trigger notifications
+FAIL test6: inline element notifies once with 0x0. assert_unreached: Timed out waiting for notification. (1000ms) Reached unreachable code
```
Why is this failing in WK1?

cathiechen replied:
updaterendering is not triggered properly in WK1, filed a bug https://bugs.webkit.org/show_bug.cgi?id=250900

Loirooriol replied:
Please link to that bug in the commit message.

cathiechen replied:
Done!

Loirooriol commented on Source/WebCore/page/ResizeObserver.cpp:
```diff
@@ -141,6 +144,10 @@ void ResizeObserver::deliverObservations()
     m_activeObservations.clear();
     auto activeObservationTargets = std::exchange(m_activeObservationTargets, { });
 
+    auto scopeExit = makeScopeExit([&] {
```
Not sure if I understand `makeScopeExit`. It seems that it will clear `m_targetsWaitingForFirstObservation` once `scopeExit` is destroyed, which I guess it's at the end of the function? But then `m_targetsWaitingForFirstObservation` will be cleared after `m_callback->handleEvent`, which seems bad, because the callbacks could be observing new elements that will be appended into `m_targetsWaitingForFirstObservation` and shouldn't be cleared yet.

cathiechen replied:
Hmm, yes, that is possible.

rniwa replied:
Why don't we std::exchange it to a local variable instead? Then the list still keeps object alive during the duration of this function, and it goes away at the end of function.
```cpp
targetsWaitingForFirstObservation = std::exchange(m_targetsWaitingForFirstObservation);
```

Loirooriol replied:
Yes, like for `IntersectionObserver` https://github.com/WebKit/WebKit/blob/fc0d8fbdae064b740a609102f1c75f9ffee73887/Source/WebCore/page/IntersectionObserver.cpp#L291

cathiechen replied:
Done

Loirooriol reviewed (commented):
If you are merging this with #8839, then I think you should use the title and bug number from that one.

rniwa reviewed (approved):
r=me

Loirooriol reviewed (changes requested):
You need to address the failure in the style bot:

> ERROR: Source/WebCore/page/ResizeObservation.cpp:44:  Code inside a namespace should not be indented.  [whitespace/indent] [4]

Loirooriol commented on Source/WebCore/page/ResizeObservation.cpp:
```diff
@@ -41,6 +41,7 @@ Ref<ResizeObservation> ResizeObservation::create(Element& target, ResizeObserver
 
 ResizeObservation::ResizeObservation(Element& element, ResizeObserverBoxOptions observedBox)
     : m_target { element }
+    , m_lastObservationSizes { LayoutSize(-1, -1), LayoutSize(-1, -1), LayoutSize(-1, -1) }
     , m_observedBox { observedBox }
```
This seems to work for `check-webkit-style`:
```suggestion
    , m_observedBox { observedBox }
    , m_lastObservationSizes { LayoutSize(-1, -1), LayoutSize(-1, -1), LayoutSize(-1, -1) }
```

rniwa replied:
I don't think style checker failing matters much.

Loirooriol replied:
Mm, yeah, I just didn't want it to start permafailing, but it seems it only checks problems in the current commit, so I guess it will be fine.

cathiechen replied:
We need to follow the order of declaration in the head file, otherwise the compiler complains.
So I'd prefer to leave as it was.

Loirooriol reviewed (commented):
The assert isn't currently failing, so I think the commit title should be about the -1x-1 to always notify observed elements.

Loirooriol reviewed (approved):
LGTM with the changes in the commit message

cathiechen committed:
[ResizeObserver] The initial last reported size of ResizeObservation should be -1x-1 which always notify observed elements
https://bugs.webkit.org/show_bug.cgi?id=251015

Reviewed by Oriol Brufau and Ryosuke Niwa.

This patch initializes the lastReportedSize of ResizeObservation to -1 x -1, which is the conclusion of [1],
it indicates that -1 x -1 is the initial size when first observing an element with ResizeObserver.

The previous change makes every ResizeObservation be delivered at least once, even if the target is a disconnected element.
When GC happens between observe and gather ResizeObservations, JSC::Weak<JSC::JSObject> m_callback inside JSCallbackDataWeak
could be released, and the element and JSResizeObserver might not be released, for they are not strong connected, this scenario
would trigger the ASSERT in ResizeObserver::deliverObservations. To fix it, this patch adds m_targetsWaitingForFirstObservation
to extend the life cycle of ResizeObserver to the first time of deliverObservations. The fix is similar to
what IntersectionObserver does in bug 231235.

The test6 of resize-observer/notify.html still fails in mac-wk1, because updaterendering is not triggered properly.
Filed [2] to track it.

[1] https://github.com/w3c/csswg-drafts/issues/3664
[2] https://bugs.webkit.org/show_bug.cgi?id=250900

* LayoutTests/imported/w3c/web-platform-tests/resize-observer/notify-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/resize-observer/observe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/resize-observer/svg-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/resize-observer/notify-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/resize-observer/notify-expected.txt.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/resize-observer/svg-expected.txt:
* Source/WebCore/page/ResizeObservation.cpp:
* Source/WebCore/page/ResizeObserver.cpp:
(WebCore::ResizeObserver::observe):
(WebCore::ResizeObserver::deliverObservations):
(WebCore::ResizeObserver::isReachableFromOpaqueRoots const):
(WebCore::ResizeObserver::removeAllTargets):
(WebCore::ResizeObserver::removeObservation):
* Source/WebCore/page/ResizeObserver.h:

Canonical link: https://commits.webkit.org/259673@main
